test/functional/autocmd/textyankpost_spec.lua

```diff
@@ -0,0 +1,209 @@
+local helpers = require('test.functional.helpers')
+local clear, eval, eq, insert = helpers.clear, helpers.eval, helpers.eq, helpers.insert
+local feed, execute, expect, command = helpers.feed, helpers.execute, helpers.expect, helpers.command
+local curbufmeths, funcs = helpers.curbufmeths, helpers.funcs
+
+describe('TextYankPost', function()
+  before_each(function()
+    clear()
+
+    -- emulate the clipboard so system clipboard isn't affected
+    execute('let &rtp = "test/functional/fixtures,".&rtp')
+
+    execute('let g:count = 0')
+    execute('autocmd TextYankPost * let g:event = copy(v:event)')
+    execute('autocmd TextYankPost * let g:count += 1')
+
+    curbufmeths.set_line_slice(0, -1, true, true, {
+      'foo\0bar',
+      'baz text',
+    })
+  end)
+
+  it('is executed after yank and handles register types', function()
+    feed('yy')
+    eq({
+      operator = 'y',
+      regcontents = { 'foo\nbar' },
+      regname = '',
+      regtype = 'V'
+    }, eval('g:event'))
+    eq(1, eval('g:count'))
+
+    -- v:event is cleared after the autocommand is done
+    eq({}, eval('v:event'))
+
+    feed('+yw')
+    eq({
+      operator = 'y',
+      regcontents = { 'baz ' },
+      regname = '',
+      regtype = 'v'
+    }, eval('g:event'))
+    eq(2, eval('g:count'))
+
+    feed('<c-v>eky')
+    eq({
+      operator = 'y',
+      regcontents = { 'foo', 'baz' },
+      regname = '',
+      regtype = "\0223"
+    }, eval('g:event'))
+    eq(3, eval('g:count'))
+  end)
+
+  it('makes v:event immutable', function()
+    feed('yy')
+    eq({
+      operator = 'y',
+      regcontents = { 'foo\nbar' },
+      regname = '',
+      regtype = 'V'
+    }, eval('g:event'))
+
+    -- the regcontents should not be changed without copy.
+    eq(pcall(command,'call extend(g:event.regcontents, ["more text"])'),false)
+
+    -- can't mutate keys inside the autocommand
+    execute('autocmd! TextYankPost * let v:event.regcontents = 0')
+    eq(pcall(command,'normal yy'),false)
+
+    -- can't add keys inside the autocommand
+    execute('autocmd! TextYankPost * let v:event.mykey = 0')
+    eq(pcall(command,'normal yy'),false)
```

Comment:
Also (or whatever error number is correct):

    neq(nil, string.find(eval('v:errmsg'), '^E46:'))